Build scripts and build system consolidation.

EraYaN opened the pull request:
Ready for testing.

Closes #331 

Requires bash, and for local builds a recent version of the dotnet sdk.

This will require users of unraid to change their template URL.

Make sure to squash and merge.

EDIT: Native dependency building should be some later PR.

nvllsvm commented:
@EraYaN - keep the `WIP: ` prefix in the title until it is ready for one of us to review and merge.

EraYaN commented:
Fixed the source package it's now under 30MB. It's ready for review.

joshuaboniface commented on deployment/clean_all.sh:
```diff
@@ -0,0 +1,16 @@
+    echo "Processing ${directory}"
+    pushd "$directory"
+    if [ -f clean.sh ]; then
+        ./clean.sh 
```
I think it should clean up the `pkg-dist` tars too, to avoid leaving cruft around. OR, the `collect_all` script should `mv` instead of `cp`.

EraYaN replied:
Agreed. Will add (in `build.common.sh`)

nvllsvm commented:
This breaks the building of the Docker containers as they expect to be executed from the root of the repo. Also, Dockerfiles cannot copy from a parent directory (ex, cannot `COPY .. .`).

joshuaboniface commented:
So, I've been looking at this more, and thinking and discussing more about it with @nvllsvm, and while we both love the idea, there's two things I want to bring up:

1. Moving the main `Dockerfile` like this actually breaks Docker builds - it expects to be in the root of the repo and won't work (due to Docker limitations) without being there. Plus, it's a good form with Docker to make `docker run .` work in the root of the repo. So we should move that, and only that, back.

2. I'm thinking about the scripts, and while they are working, scripting everything like this in Bash is both potentially non-portable (as we discussed earlier on Riot regarding bash versus other shells) and fragile - we have to keep all these scripts up-to-date, worry about consistent versioning, etc. And while there are tools to help with this, I think it might be easier to go a different direction: GNU Make.

Make should be able to do everything we want with regards to consistent "execute these build tasks, then these packaging tasks", handling version logic, and other tasks, while avoiding repeating code and keeping all the per-package-type configs separate from the main repo dir. And this lets us maintain just a single `Makefile` in the root of the repo that can be used easily without traversing directories, which is going to be a barrier to new users.

So what I'd like to do is rebase this PR against a new feature branch, `packaging-improvements`, where we can work on this in isolation without disrupting `dev`, and can then work towards this goal. My endgoal would be to be able to run `make deb` or `make win` or `make all` from the root of the repo, and end up with the package(s), named properly, in the parent dir. And I would still like @LogicalPhallacy to comment on how this would affect Windows builds, though I forsee them working well the same way.

I'll leave this as-is for tonight, but please let me know your thoughts too, and I'll do the branch related stuff tomorrow.

sparky8251 commented on deployment/debian-x64/package.sh:
```diff
@@ -0,0 +1,29 @@
+mkdir -p "$package_temporary_dir"
+mkdir -p "$output_dir"
+docker run --rm -v "$package_temporary_dir:/temp" "$image_name" sh -c 'find / -maxdepth 1 -type f -name "jellyfin*" -exec mv {} /temp \;'
+chown -R "$current_user" "$package_temporary_dir"
```
Should you not instead add a bit at the beginning requesting root if it doesn't have it already? I personally don't like the idea of sudo's in a script. I might have used it recently and now its going to run a script with my cached creds, potentially doing something I don't want.

I have a script I wrote for another project that handled this at least in part. Can see it here: https://github.com/morph1904/TygerCaddy/blob/master/tyger-install-ubuntu.sh

EraYaN replied:
It should all write to user writable folders. So sudo should not be required. Since both $output_dir and $package_temporary_dir should be within the source tree. (Which you should have write access to.

Sudo in a script is a bad idea IMO.

sparky8251 replied:
`docker` itself requires root access for *all* commands. You can add yourself to the docker group, but thats basically the same as granting yourself root access to the entire system since you can mount system files as volumes and edit them in a container.

It's best to assume that admins and security conscious folks will *not* be adding themselves to the docker group for the above reason.

EraYaN replied:
On any of the default installs of docker on dev machines, this works just like it is. Both Docker for Windows (with the client in WSL) and Docker for Mac, handle this out of the box. I suppose one could run the whole script as root.

sparky8251 replied:
Ah, well... I know at least @joshuaboniface and I use Linux as our dev machine and by default, Linux requires root access.

joshuaboniface replied:
Itt needs to be sudo'd because the file moved from the `docker` container ends up owned `root:root`, so the unprivileged user running the command has no perms to edit (and hence `mv`) it. Or chown has to happen in docker but that might not work because the user doesn't exist in the container.

EraYaN replied:
Added a condition if the non-sudo fails it will retry with sudo.

EraYaN commented on deployment/collect_all.sh:
```diff
@@ -0,0 +1,20 @@
+    echo "Collecting everything from '$directory'.."
+    PLATFORM=$(basename "$(dirname "$directory")")
+    # Copy all artifacts with extensions tar.gz, deb, exe, zip, rpm and add the platform name to resolve any duplicates.
+    find $directory \( -name "jellyfin*.tar.gz" -o -name "jellyfin*.deb" -o -name "jellyfin*.rpm" -o -name "jellyfin*.zip" -o -name "jellyfin*.exe" \) -exec sh -c 'cp "$1" "'${COLLECT_DIR}'/jellyfin_'${PLATFORM}'_${1#*jellyfin_}"' _ {} \;
```
The tar.gz archive it copies is the source package. This is only for uploading, so I thought the *.dsc and others should not be included. For publishing to a debian repo, I think there is an other tool. `dput` or something.

EraYaN commented:
_About docker:_ 
It's not required anymore. WE can always have a single dicker-run.sh script in the docker folde rin deployments. Especially because the root of the repo is already cluttered with all the source directly there instead on in a Solution folder or `src` folder. Since Docker is not a main target, I personally hate to bend the structure of the project to Docker. People agreed, hence the PR linked below.
https://github.com/docker/cli/pull/886 (stable had this since last April 24).

The docker correct command would be (it has the build-context at the root, just the Dockerfile is not there)
`docker run . -f deployment/docker/Dockerfile`
or for the other arch
`docker run . -f deployment/docker/Dockerfile.aarch64`

_About make:_
I like it, it's mostly bash-like anyway. But the Riot chat was very against it. Since it is a build system `make` on top of a build system `dotnet`. And bash scripts are easier to maintain IMO, especially for less experienced Makefile writers. I have written many a unmaintainable mess in Makefile format before ;). Especially packaging in Makefiles is quite horrible.

And yes the scripts are a bit fragile, because self location detection is even more fragile, so they assume a certain `pwd` (which would be their own directory)

I would love to just have it all in python or somethng. That way we can update alll the changelogs for all packages automatically and manage all other versions directly from the AssemblyVersion. (And also tag everyone with the git sha1 (short hash). Or update it into `SharedVersion.cs`. So every class can access it as well. So many possibilities. The only platform that does not ship with python3 is Windows.

Lastly these scripts are not for development, these are the "code-frozen"->"build_all"->"publish_all"->"collect_all"->"github release upload"
All in the name of convenience and repeatability. There is still quite some code repetition, but that is because different platforms might require different things.
And the more is automated the better.

Also, the debian source package right now ignores the `ThirdParty` directory. Is this correct? Do we package the source of dependencies? It does stop the source package from building though.

EraYaN commented:
I started out on Makefiles (see first couple of revisions of the OP). But the room was not about that. And I sort of understood, so I went for sh first. And then later bash.

If you search for `"My thoughts: NO MAKE"` in the #jellyfin-dev channel. You will see the conversation from yesterday about it.

joshuaboniface commented:
Found it! @Bond_009 why no make?

joshuaboniface commented:
Found it! @Bond-009 why no make?

LogicalPhallacy commented:
So while building all from windows is unlikely to be a priority, the build all script should probably build windows 32bit as well. Additionally windows presents the challenge of needing ffmpeg binaries (and probably the NSSM binaries as well). Both can be bundled in, that's not really a concern, but you'll want to take that bundling code from build-jellyfin.ps1 and include it (FFMPEG is GPL, NSSM is public domain so the bundle is a nonissue AFAIK). I would suggest moving build-jellyfin.ps1 to a "win-generic" folder as well, since technically windows CAN run on ARM and ARM64 and in the future that might be relevant if FFMPEG builds for those architectures become acailable.

LogicalPhallacy commented:
>           Are there still x86 only windows'es? ffmpeg on 32 bit is horribly slow (x264/x265 encoding especially). Not sure we should support that.
> macOS should also have ffmpeg bundled in the app bundle. But the same thing as for windows applies, that is something for the setup to do. The portable version should probably be as lean as possible.

WIndows 10 Client and Server 2008 both support x86 mode. While its uncommon, there are some users with legacy server hardware that might appreciate it, and the effort to build it is minimal.  As for the setup bundling rather than the build I generally agree, but one of the things that helps draw new users is ease of setup, bundling into DMG (for MacOS) or exe/appx/msi installers (for windows) is a ways out, until then fetching the binaries needed at build time to packaged in a zip seems the least we can do.

EraYaN commented:
Rebased and updated according to comments. It will only try sudo if the previous command fails. (So on macOS it can still run non-interactively.

EraYaN commented:
It technically is make without make, but with added `if` I suppose.

joshuaboniface commented:
So I think for now, I'd like to go ahead with this, assuming @nvllsvm 's concerns are handled - I think Docker is weird and different enough from the other build types that the `Dockerfile` can just live in the root. But for the rest of it, if we're going to keep packaging integrated it's gotta be well-organized like this, and we can add codeowners for the various parts of it including their small `bash` scripts to ensure that package maintainers are selected.

I've reconsidered my Make suggestion due to the `if` issue. I think `bash` is the next best thing, and I've come to really like the separate folder scripts due to the reasons above. It's really consistent and can be handled differently for each package type giving us a lot of flexibility.

Plus, `build all` will be amazing for automating the builds eventually.

I have two major suggestions from myself:

1. A nicer central script that can exist in the repository root called `build.sh` that takes flags like `all` or an individual type (folder) name and handles collect and package too. I can write this fairly easily too if you wish and add it "later" (i.e. right after this is merged).

2. Some more obvious organization, clear separation between building standalone binaries and packages, including making the Linux binaries generic (as they should be). I'm thinking something like:

```
debian-package-x64/
debian-package-x64/pkg-src
debian-package-x64/build.sh
...
linux-x64/
linux-x64/build.sh
...
docker/
...
unraid-docker/
...
```

Also what is `framework`? I'm not sure what that one does. 

Anyways I'm open to other thoughts on this, but as mentioned am pretty convinced with this solution after using it for a while.

joshuaboniface commented on deployment/debian-x64/package.sh:
```diff
@@ -0,0 +1,33 @@
+mkdir -p "$package_temporary_dir"
+mkdir -p "$output_dir"
+docker run --rm -v "$package_temporary_dir:/temp" "$image_name" sh -c 'find / -maxdepth 1 -type f -name "jellyfin*" -exec mv {} /temp \;'
+chown -R "$current_user" "$package_temporary_dir"
```
Actually, I'd also like it to not do such a broad find on `/` in the container.

I think this is a consequence of trying to use this Dockerfile to do the generic binary as well. As mentioned in my previous comment, we should split that out, so there wouldn't be any need to alter the flow from how the original `build-deb.sh` script does it, i.e. copying just the files from `/dist`. For now this works though, and I'll do modifications to this later too.

EraYaN replied:
Well it only copied from /dist because it hard codes a move of just the binary package. Since we also want the source package and desc files we need the find. It's not all that broad since the depth is limited to 1.

I guess I can move the find to the Dockerfile. But it does not make a lot of difference. If you do not want in in the `/` directory the Dockerfile should add an extra layer of directory to `/repo` since the deb will ALWAYS land in `..` this is non configurable (due to all the weird cross dependencies in the tooling.

But in the end it's semantics. I can change it either way.

joshuaboniface replied:
I'm not sure I understand why the `find` is needed, is it not putting all the `dpkg-buildpackage` output into a single place that can be copied as a whole?

joshuaboniface commented on deployment/debian-x64/Dockerfile:
```diff
@@ -11,12 +13,11 @@ RUN apt-get update \
- && mkdir /dist \
- && mv /jellyfin*deb /dist
+RUN yes | mk-build-deps -i debian/control \
+    && dpkg-buildpackage -us -uc
```
Another case for splitting out the binary build from the package build: it's actually totally redundant to do ` dotnet publish` and copy the files into the docker image, only to run `dpkg-buildpackage` which rebuilds it (and *must* do so).

EraYaN replied:
Not everyone might want a *.deb package though. I mean I guess they could then grab the linux-x64 package?

joshuaboniface replied:
Right but we need a specific build setup for the Deb package in particular, separate from the others, so I can build it consistently. I don't like the overlap is all, I'd rather have a generic Linux binary, then a Debian package that does zero `dotnet` work on the current machine (only in Docker like the current script).

joshuaboniface replied:
Same idea when we integrate the RPM package, it would have a totally separate folder and a totally separate build process, also inside Docker.

EraYaN replied:
Well the debian package is completely separate from the portable debian build, they are just in the same script after one another. I can pull those apart if you want.

joshuaboniface replied:
Yes please, let's keep it totally separate: `debian-package-x64` folder that only does the packages (src + binary), then a `debian-x64` that does the portable build.

joshuaboniface commented on deployment/linux-x64/build.sh:
```diff
@@ -0,0 +1,7 @@
+
+VERSION=`get_version ../..`
+
+build_jellyfin ../.. Release linux-x64 `pwd`/dist/jellyfin_${VERSION}
```
Please add a trailing newline.

joshuaboniface commented on deployment/ubuntu-x64/build.sh:
```diff
@@ -0,0 +1,7 @@
+
+VERSION=`get_version ../..`
+
+build_jellyfin ../.. Release ubuntu-x64 `pwd`/dist/jellyfin_${VERSION}
```
Please add a trailing newline.

joshuaboniface commented on deployment/win-x86/package.sh:
```diff
@@ -0,0 +1,9 @@
+
+package_portable ../.. `pwd`/dist/jellyfin_${VERSION}
+
+#TODO setup and maybe change above code to produce the Windows native zip format.
```
`zip` can be done on Linux easily, and the existing ZIP files from @LogicalPhallacy just put all the files in there (though I'd prefer a subdir for consistency with most `zip`'d programs).

EraYaN commented:
This should fix the taglib-sharp build issue as well. (Target framework)

This is broken with the `dotnet` CLI. MSBuild/VS doesn't care.

Also should integrate #420

EraYaN commented:
Framework is the non specific build that generates a *.dll to be run with `dotnet` CLI. It is the final catch all.

anthonylavado commented:
We might need to push this slightly to the first maintenance of 10.x, so 10.0.1. This will probably be over the coming week.

This is one we want to get right, 100%.

EraYaN commented:
And we need to include native builds for libSkiaSharp for ARM integrated in the build system (#335)

Here is the upstream issue tracking this: https://github.com/mono/SkiaSharp/issues/453#issuecomment-450497754 The linked comment provides cross-building instructions.

EraYaN commented:
Also incorporates #476.

EraYaN commented:
I added the main build.sh, it takes an array of commands (or all) with the folder names. a translation can be done later by some bash scripting guru.

Also I found the official workaround for the tablig-sharp building. We shouldn't try to publish the solution file, because it's pretty broken. It's also faster to just publish Jellyfin.Server.17:33
Source: https://github.com/dotnet/cli/issues/9149#issuecomment-430366906

@joshuaboniface Please check my debian package build rules. (Added the last positional argument "Jellyfin.Server")

joshuaboniface commented:
@EraYaN Fantastic, thanks! Gonna test this out tomorrow and let you know how it looks.

joshuaboniface commented:
@nvllsvm Just need your last thoughs on this one - I think moving the Docker, now that we also have an Arm one, makes more sense. Maybe we can add a `docker-run` script to the root afterwards to auto-detect the arch and run it without needing the full path? In any case, I'd rather get these changes in-tree sooner rather than later, and then modify what's needed from there, rather than leave it lingering.

joshuaboniface commented:
@Wuerfelbecher We can get your changes in here too, either you rebase after or @EraYaN can integrate the changes. I implicitly approve that PR as part of this, and we can clean up some of the duplicate files in-tree afterwards.

EraYaN commented:
I merged in the RPM changes, but it seems to build twice. It will remove the `dotnet build` command.

nvllsvm commented:
I do not like having the Dockerfile outside of the root directory. It might be ugly, but it's dead-simple for one who works with Docker. As of now, I reject the change to move it elsewhere, but am open to more convincing.

EraYaN commented:
There is really no reason to have any of the Dockerfiles in the root. The build will be automated and for users they can use the dockerhub images. So now nothing clutters the root of the repository and is all neatly packaged and hidden away. This also gives us the possibility to move all the files for building/packaging/deploying/whathaveyou to a seperate repo altogether. Nobody really should ever have to type in the command to build for docker. I suppose if you really want to, the command would be along the lines of `cd deployment && ./build_all.sh docker`. And that should build all relevant docker images and create the manifest. If we split out the UI completely I suppose we could even move to `docker-compose.yml` (if we split the docker images as well).

The docker images will be linked together into a manifest and that can be uploaded to dockerhub. Right now it's `package.sh` although the final publish command is disabled. Manifests seem to not be able to be created locally. (Manifests are those multi-arch images)

I suppose I can add a symlink for the Dockerfile into the root? But that is still just clutter IMO.

Besides all that, we're also figuring out an issue with the rpm-build adding liblttng-ust as a dep. While jellyfin really should not require that. (and runs fine when you pass `--nodeps` to the rpm command). So I will probably add a `AutoReq: no` line to the *.spec file, testing right now. And will push my remaining changes when done.

Lastly I have updated the arm docker image to be more explicit in their arm source image choice, but we might need to add a different repo to install ffmpeg 4.0.3, right now it seems to install the version 3.2.x.

EraYaN commented:
So the only thing to do still would be the `sign.sh` & `publish.sh` scripts. And some opts parsing the the main `make.sh`.

nvllsvm commented:
Thanks for explaining your reasoning. I'd still like them to be in the root directory. To me, I enjoy the clarity and convenience of seeing a Dockerfile in the root of a project. I know that running `docker build .` should be most or all of what I need to do.

Don't split the arm platform images. It unnecessarily duplicates code and Docker is smart enough to pick the correct platform. Instead, rename `Dockerfile.aarch64` to `Dockerfile.arm`. 

Also, keep the original `Dockerfile` as-is - not renaming to `Dockerfile.amd64`. It should be considered the permanent Dockerfile with the arm variant being temporary until .NET Core 3.0 is stable.

I still need time to the review the rest of the changes.

EraYaN commented:
Since we have a native dependency on ffmpeg, which we will need to switch out for a self built one, so for that we will need the separate images. Hence why I split them. An other way is to create a base image (containing the correct ffmpeg build) and create the correct manifest for that one. But really that just adds another layer, doesn’t really make it any ‘better’ in any way.

Wuerfelbecher commented:
> rpmlint still gives a ton of warnings and errors but alas. I'm not hoe enough in the fedora world to make sense of those error.

This is the output i got from `rpmlint`:

```shell
# rpmlint jellyfin.spec
jellyfin.spec:71: E: hardcoded-library-path in %{_prefix}/lib/firewalld/service/%{name}.xml
jellyfin.spec:87: E: hardcoded-library-path in %{_prefix}/lib/firewalld/service/%{name}.xml
jellyfin.spec: W: invalid-url Source0: jellyfin-10.0.1.tar.gz
0 packages and 1 specfiles checked; 2 errors, 1 warnings.
```

`rpmlint` will throw errors because of the firewalld  service `.xml` since the path is kind of hard coded. I didn't find a RPM macro that resolves to `/usr/lib/firewalld/service` where service definitions belong for firewalld and since `%{_lib}` resolves to `lib64` (but only on 64-bit systems), i went with `%{_prefix}/lib/firewalld/service` so `rpmlint` obviously doesn't like that but `rpmbuild` will build an RPM anyway. 

Warnings about invalid Source urls can be safely ignored since if the specified there is not present when rpmlint is run it will complain but as far as i understand the `package.sh` script it will create a source tarball and if you run rpmlint again with the tarball present that line will disappear. 

@EraYaN I hope that i explained the rpmlint errors and the reasoning behind the somewhat hard coded firewalld service path.

EraYaN commented:
Alright, cool, I might have made a mistake in running my rpmlint. Cause I got like 100+ warnings and errors. Ooh well. I tested the software at it works so yay.

EraYaN commented:
@nvllsvm Moved the main Dockerfile to the root.

Before running docker build one will need to execute both the submodule command and later the version command.

JustAMan commented on deployment/docker/Dockerfile.arm64v8:
```diff
@@ -0,0 +1,17 @@
+WORKDIR /repo
+COPY . .
+RUN export DOTNET_CLI_TELEMETRY_OPTOUT=1 \
+ && find . -type f -exec sed -i 's/netcoreapp2.1/netcoreapp3.0/g' {} \; \
```
this magic needs a `TODO` or something so we won't forget to either update it when we move to `netcore 2.2` or remove when we move to `3.0` (or when MS would release .net 2.1 for arm64).

EraYaN replied:
This should probably be taken care of by drone builds for all platforms.
I will add a TODO as well.

nvllsvm commented:
@EraYaN - Move the arm dockerfiles to the root as well

Bond-009 commented on .gitignore:
```diff
@@ -234,23 +233,33 @@ pip-log.txt
+##########
+# Visual Studio Code
+##########
+.vscode/
```
This is already in the repo, @anthonylavado

anthonylavado replied:
I don't remember commenting about it, but it's good that it's there :sweat_smile:

brianjmurrell commented:
Is the way this is organised for Fedora actually going to be buildable in [COPR](https://copr.fedorainfracloud.org/)?

Leveraging native package building systems (i.e. COPR for Fedora and EL, PPA for Ubuntu, etc) really ought to be a primary goal when organising builds for O/Ses that have package building systems available.

Wuerfelbecher commented:
For manual builds COPR only cares about the SRPM but for automatic builds like with the github integration it needs either all the required sources from the `.spec` already in the directory containing the spec (at least for the `rpkg` build method) or a [Makefile that outputs an SRPM](https://docs.pagure.org/copr.copr/user_documentation.html#make-srpm).

COPR builds would work with a small Makefile that just creates the tarball and builds the SRPM with something like `rpmbuild -bs jellyfin.spec --define "_sourcedir $PWD" --define "_srcrpmdir $outdir"`.

COPR builds could cover CentOS, Fedora 28/29/rawhide and possibly also OpenSUSE.

Wuerfelbecher commented:
@EraYaN can you also add the directory for the CentOS Package the `.spec` and package-sources are identical and can be symlinked but building is different since CentOS only has ancient packages.

This Dockerfile should work:
```dockerfile
FROM centos:7
ARG HOME=/build
RUN mkdir /build && \
    yum install -y @buildsys-build rpmdevtools yum-plugins-core && \
    rpm -Uvh https://packages.microsoft.com/config/rhel/7/packages-microsoft-prod.rpm && \
    rpmdev-setuptree

WORKDIR /build/rpmbuild
COPY ./deployment/centos-package-x64/pkg-src/jellyfin.spec SPECS
COPY ./deployment/centos-package-x64/pkg-src/ SOURCES

RUN spectool -g -R SPECS/jellyfin.spec && \
    rpmbuild -bs SPECS/jellyfin.spec && \
    yum-builddep -y SRPMS/jellyfin-*.src.rpm && \
    rpmbuild -bb SPECS/jellyfin.spec;
```

brianjmurrell commented:
@Wuerfelbecher Indeed!  I would advocate for the Makefile and the use of webhooks (i.e. on tag -- for the release COPR build at least) so that COPR building was automatic on tagging.
